### PR TITLE
fix: dont force exit when it was explicitly disabled

### PR DESCRIPTION
**Pull request by pimlie**
fix: remove slash from warning text
fix: dont force-exit when explicitly disabled
chore: add force-exit tests for command

Follow-up of: #4958 

## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

As discussed on discord, some fixes and stability improvements for my pr from yesterday

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/cli/src/command.js [hunk 1]
```diff
@@ -48,9 +48,12 @@ export default class NuxtCommand {
 
     const runResolve = Promise.resolve(this.cmd.run(this))
 
-    // TODO: For v3 set timeout to 0 when force-exit === true
-    if (!this.isServer || this.argv['force-exit']) {
-      runResolve.then(() => forceExit(this.cmd.name, forceExitTimeout))
+    if (
+      this.argv['force-exit'] ||
+      // dont force exit when it was explicitly disabled by the user
+      (!this.isServer && !this.isExplicitArgument('force-exit'))
```
Comment (manniL) [inline packages/cli/src/command.js]: Why `!this.isServer`? 🤔

File: packages/cli/src/options/common.js [hunk 2]
```diff
@@ -28,7 +28,6 @@ export default {
       }
     }
   },
-  // TODO: Change this to default: true in Nuxt 3
```
Comment (manniL) [inline packages/cli/src/options/common.js]: Why removing that todo? 🤔

File: packages/cli/src/command.js (hunk 1 above)
Comment (pimlie) [reply packages/cli/src/command.js]: Its to differentiate between build/generate and dev/start without having the need to change the default value of force-exit

**Comment by pimlie** [comment]
As there appear to be some confusion (not in the least by myself), here is to clarify the force-exit behaviour:

- We want nuxt-cli to be able to forcibly exit after `NuxtCommand.run()` has finished. This behaviour can be toggled on the cli by using `--force-exit` and is _independent_ of the actual command that is run (its an abstract feature).
That means you can also run `nuxt start --force-exit` and then nuxt will immediately exit after it started listening. Maybe not really usefull normally but it could be used to quickly or automatically test whether the Nuxt server wa setup correctly and didnt threw any errors on server middleware etc.

- If you use `--force-exit` then process.exit will _always_ be called with exitCode `0`. This is because this way you have a way to ignore bad code and force eg your ci pipeline to continue. Besides, when there are fatal errors then nuxt will already exit with a non-zero exit code in the consola reporter so we dont have to check that here <sub>(but we might want to anyway in the future for non-fatal errors)</sub>.

- For some commands we want an additional check, `nuxt build` and `nuxt generate` are commands of which we know they should immediately exit after `NuxtCommand.run()` has finished. For this we added a (reversed) additional check on `this.isServer`. isServer means we _assume_ the command is starting a server and its normal the command keeps running after `NuxtCommand.run()` finishes. So the reverse is that when `!isServer` it shouldnt keep running so we only force-exit when `bin/nuxt-cli.js` doesnt exit by itself.

- For this additional check we _dont_ always want to call process.exit as it would hide bad code. So instead of calling process.exit immediately we set an unref timeout that calls process.exit. By unreferencing the timeout node will just normally exit when the event loop is empty, but if there are unreleased listeners still using the event loop the timeout will trigger after 5 seconds and still exit nuxt.

The difference between nuxt v2 and v3 will be that the additional check for `nuxt build/generate` is a warning and exits with code 0 in v2 while in v3 it will be a (fatal?) error and will exit with non-zero code.

File: packages/cli/src/command.js [hunk 3]
```diff
@@ -102,6 +105,10 @@ export default class NuxtCommand {
     return new Generator(nuxt, builder)
   }
 
+  isExplicitArgument(option) {
```
Comment (pi0) [inline packages/cli/src/command.js]: Can be better to name it `hasExplicitArg` ?

Comment (pi0) [reply packages/cli/src/command.js]: or `argDisabled` or something more meaningful :)

**Comment by pi0** [comment]
> * That means you can also run `nuxt start --force-exit` and then nuxt will immediately exit after it started listening. Maybe not really usefull normally but it could be used to quickly or automatically test whether the Nuxt server wa setup correctly and didnt threw any errors on server middleware etc.

This throws inappreciate warning/error message.

> * For some commands we want an additional check, `nuxt build` and `nuxt generate` are commands of which we know they should immediately exit after `NuxtCommand.run()` has finished. For this we added a (reversed) additional check on `this.isServer`

Two remarks:
- `isServer` check is not accurate. We rely on an implicit option to determine it.
- I think it could be better to explicitly enable `force-exit` by default in their cli options. And probably allow `--no-force-exit` as well.

**Comment by pimlie** [comment]
> This throws inappreciate warning/error message.

I dont think it does? As mentioned above when you explicitly set `--force-exit` I dont set the timeout and you dont get the warning message. See here: https://github.com/nuxt/nuxt.js/blob/e84af4b79d289c9de8ab0ba196e02727d27fc5ef/packages/cli/src/utils/index.js#L143

> * `isServer` check is not accurate. We rely on an implicit option to determine it.
> * I think it could be better to explicitly enable `force-exit` by default in their cli options. And probably allow `--no-force-exit` as well.

By using `isServer` I wanted to differentiate between the two options, because using `--force-exit` should be _always call process.exit when cmd.run finishes_ and the additional check is _only force exit when nuxt-cli doesnt exit by itself_. It doesnt feel correct to me to use `--force-exit` for both situations.

Maybe we should otherwise add a cmd property?
```js
export default {
name: 'dev',
isServer: true,
run() ...
}
```

**Comment by pi0** [comment]
@pimlie Exactly. I suggest allowing setting forceExit in `export default` of `build` and `generate` command that explicitly needs them. Less config and best behavior by default. Use can disable them if rare conditions by `--no-force-exit` them. And no more relies on `isServer`.

**Comment by pimlie** [comment]
I added the possibility of making the default option value a fn. Good or not? Description is now a bit less descriptive because its the same description for both `--force-exit` as `--no-force-exit`.

File: packages/cli/src/options/common.js [hunk 4]
```diff
@@ -30,8 +30,10 @@ export default {
   },
   'force-exit': {
     type: 'boolean',
-    default: false,
-    description: 'Force Nuxt.js to exit after the command has finished (this option has no effect on commands which start a server)'
+    default(cmd) {
+      return ['build', 'generate'].includes(cmd.name)
```
Comment (pi0) [inline packages/cli/src/options/common.js]: Good trick :+1:

**Comment by pimlie** [comment]
Another day and another night thinking about this in bed. Or should we just leave the default option value and its description as is/was and do this:

```js
const cmdShouldFinish = ['build', 'generate'].includes(this.cmd.name)
const forceExitByUser = this.isUserSuppliedArg('force-exit')
if (this.argv['force-exit'] || (cmdShouldFinish && !forceExitByUser)) {
  runResolve.then(() => forceExit(this.cmd.name, forceExitByUser ? false : forceExitTimeout))
}
```

So that would actually be still some sort of a `isServer` variable but instead of looking at options.hostname it just checks the cmd.name. I like this better I think

**Review by pi0 — APPROVED** [review]

**Pull request merged**